v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/DataSourceProvider.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.source;
+
+import static com.google.cloud.teleport.v2.utils.KMSUtils.maybeDecrypt;
+
+import com.google.cloud.teleport.v2.options.SourceDbToSpannerOptions;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import javax.sql.DataSource;
+import org.apache.beam.sdk.io.jdbc.JdbcIO;
+import org.apache.beam.sdk.options.ValueProvider.StaticValueProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DataSourceProvider
+    implements org.apache.beam.sdk.transforms.SerializableFunction<Void, DataSource> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DataSourceProvider.class);
+  private static volatile DataSource dataSource = null;
+  // VarHandle provides a strongly typed reference to @dataSource
+  // and supports various atomic access modes like the acquire-release semantics.
+  private static final VarHandle DATA_SOURCE;
+  private final JdbcIO.DataSourceConfiguration config;
+
+  // Bind DATA_SOURCE to dataSource
+  static {
+    try {
+      MethodHandles.Lookup lookup = MethodHandles.lookup();
+      DATA_SOURCE =
+          lookup
+              .in(DataSourceProvider.class)
+              .findStaticVarHandle(DataSourceProvider.class, "dataSource", DataSource.class);
+    } catch (ReflectiveOperationException e) {
+      // Logger might not be initialized in static scope!
+      System.err.println("Error while binding VarHandle: " + e.toString());
```

Comment:
Besides showing the error, is there any actionable item the users could try to resolve this?

Reply:
I would understand the fact that it may emit error as the initialization timing is not quite accurate. It is a static member but its accesses, besides initialization are all in non-static context.

Is this a connection pool pattern (all DataSourceProvider instances share the same DATA_SOURCE)? If so, this static member should be acceesses when first DataSourceProvider constructor gets initialized or before first access, instead in static context

Reply:
I do not see the need of use of reflection here. why not just use dataSource?

Reply:
We use `DATA_SOURCE` varHandle to access `dataSource`  through relaxed atomic access. Ref: [Double-checked_locking for Java9](https://en.wikipedia.org/wiki/Double-checked_locking#:~:text=Java%209%20introduced%20the%20VarHandle,synchronization%20order%2C%20the%20global%20order)

The error to bind varHandle should not occur unless the code is modified and any of the type casting fails. Since the error occurs at static initialization the code developer should be able to catch it as a part of test flow and fix it.

Reply:
lgtm as it's a well practiced pattern, good to learn, sorry for confusion